### ISSUE TITLE
`lerna bootstrap` v3.4.3 no longer sets `process.env.npm_package_engines_*`...

### ISSUE THREAD

**Issue by busticated** [lerna/lerna#1752]
hey there :wave:

i have a monorepo that i use to manage a handful of vscode extensions. i have scripts set up like so:

in my top-level `package.json` file:

```javascript
"private": true,
"scripts": {
    "postinstall": "npm run bootstrap",
    "bootstrap": "lerna bootstrap",
    ...
},
"devDependencies": {
    "vscode": "^1.1.21",
    ...
}
```

then each of my packages use a `package.json` like:

```javascript
"private": true,
"scripts": {
    "postinstall": "node ../../node_modules/vscode/bin/install",
    ...
},
"engines": {
    "vscode": "^1.23.0",
    ...
}
```

the `vscode/bin/install` program expects to find a value for `process.env.npm_package_engines_vscode` ([source](https://github.com/Microsoft/vscode-extension-vscode/blob/master/bin/install#L12)) but when i run `npm i` at the top-level of my monorepo, i get:

```
Error installing vscode.d.ts: Missing VSCode engine declaration in package.json.
```

inspecting `process.env`, i see:

```
INIT_CWD: '/path/to/my-repo',
PWD: '/path/to/my-repo/packages/myext-vscode-core',
npm_package_name: 'myext-vscode-core',
...
```

but no trace of anything using the `npm_package_engines_` prefix.

when i `cd ./packages/myext-vscode-core` then run `npm i` and inspect `process.env` there, i see:

```
INIT_CWD: '/path/to/my-repo/packages/myext-vscode-core',
PWD: '/path/to/my-repo/packages/myext-vscode-core',
npm_package_name: 'myext-vscode-core',
npm_package_engines_vscode: '^1.23.0',
...
```

which is to say: env vars using the `npm_package_engines_` prefix are available.

all of this worked fine under lerna v2.11.0 - running `npm i` at the top-level executed my `bootstrap` script which in turn ran my packages' `postinstall` scripts without issue. i never saw the `Missing VSCode engine` error.

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.4.3 |
| `npm --version`  | 5.10.0 |
| `yarn --version` | N/A |
| `node --version` | 8.12.0 |

| OS | Version |
| --- | --- |
| macOS High Sierra | 10.13.6 |

**Comment by busticated** [comment]
fwiw, running `lerna exec --no-bail -- npm run postinstall` produces the expected results:

```
Detected VS Code engine version: ^1.23.0
Found minimal version that qualifies engine range: 1.23.0
Fetching vscode.d.ts from: https://raw.githubusercontent.com/Microsoft/vscode/7c7da59c2333a1306c41e6e7b68d7f0caa7b3d45/src/vs/vscode.d.ts
vscode.d.ts successfully installed!
```

**Comment by ThisIsMissEm** [comment]
Does this work in any of the lerna 3.x.x series?

**Comment by busticated** [comment]
I haven’t tried but I can if that’ll be helpful 👍

**Comment by busticated** [comment]
@ThisIsMissEm ok, here's what i found:

v3.0.0 - `lerna bootstrap` at the top-level does not run `postinstall` in packages, `lerna run postinstall` doesn't appear to work
v3.1.0 - same behavior as `v3.0.0`
v3.2.0 - `process.env.npm_package_engines_vscode` is NOT set when running `lerna bootstrap`, `lerna run postinstall` works ([yields this result](https://github.com/lerna/lerna/issues/1752#issuecomment-434772659))
v3.3.0 - same behavior as `v3.2.0`
v3.4.0 - same behavior as `v3.2.0`
v3.4.3 - same behavior as `v3.2.0`

**Comment by evocateur** [comment]
`lerna bootstrap` is basically legacy code at this point, I haven't personally used it in over a year. I'd say the `lerna run postinstall` route is the correct one.

**Comment by evocateur** [comment]
```sh
# run once
npx lerna link convert

# thereafter, in the root
npm install

# or `npm ci`, the whole tree is locked
```

**Comment by busticated** [comment]
does that hoist everything to the root? honestly confused on what the”happy path” is these days. 

my packages require non-dev deps to be installed locally (not hoisted to the root) since vscode essentially zips up node_modules and includes them in the published extension.

as it is today using learna v2, I’m able to install all root and package deps via `npm I` at the top-level which also runs postinstall routines in each package. this bug was filed b/c this no longer works in V3. happy to do all of this using new / different lerna features but I don’t know what those might be

**Comment by busticated** [comment]
> I'm not sure I understand why you're even using lerna for what sounds like something that very much wants to be its own repository.

all dev deps are shared, vscode extensions ship individually and together as a “pack”

**Comment by ThisIsMissEm** [comment]
Why not create a "@my/vscode-dev-packages" package which depends on all the shared packages?

**Comment by busticated** [comment]
to be clear, you don't owe me anything. i'm _not_ entitled to your free labor. i'm thankful that this helpful tool has been made freely available to me and the broader community. further, as even just a casual observer i've been impressed by how successful you all have been lately in cleaning up some of the bit rot / legacy weirdness that had built up over the years here. i know how _hard_ that work is so again, thank you :pray:

for my issue, there are likely a number of work-arounds i can explore - including simply sticking w/ v2 for now. i'd be interested in pointers on where i might dive in as well as learning more about the technical hurdles / risks around fixing my issue. naively, it seems since `lerna run postinstall` does the right thing, getting `lerna bootstrap` to do the same would be fairly straightforward but what do i know 🤷‍♂️ 

as to the notion that `lerna bootstrap` is legacy to be avoided and that my use-case is outside what lerna was designed for, i can't help but quote from the [README](https://github.com/lerna/lerna/blob/9f5c824b2f15119c10003dcc709f3ca0da99a5fe/README.md)

> Splitting up large codebases into separate independently versioned packages is extremely useful for code sharing. However, making changes across many repositories is messy and difficult to track, and testing across repositories gets complicated really fast.

this is exactly why i opted to use lerna and to date it has served me well 👍 more specifically, it allows me to share all my _dev_ dependencies and test tooling (e.g. fixtures, linting, etc), as well as providing easy publishing and orchestration (via `lerna run` and `lerna exec`) of my packages which are versioned together. if there are specific gotchas / "you might have a bad time if" type things here, i (and i imagine most everyone) would appreciate seeing them called out.

> What can Lerna do?
The two primary commands in Lerna are `lerna bootstrap` and `lerna publish`.
`bootstrap` will link dependencies in the repo together. `publish` will help publish any updated packages.

🤔🤷‍♂️


🙏 ❤️ 👍

**Comment by evocateur** [comment]
Thanks for being patient with me. :)

On further reflection, I had a hunch it might be related to the use of `npm-lifecycle`, which we now use instead of allowing the subprocess `npm install` to run the scripts. (Spoilers: yes, but not where I expected)

At first, I thought it was a bug in the library, but just now I read the source, and `npm-lifecycle` is indeed [creating those env vars](https://github.com/npm/npm-lifecycle/blob/d7a014f9393da76ee0ecf6701be84ed1ea613b83/index.js#L361) (instead of what I dreaded, something leftover in the npm CLI). So it's an issue with the `pkg` parameter (`data` in the `makeEnv()` function).

[Turns out](https://github.com/lerna/lerna/blob/9f5c824b2f15119c10003dcc709f3ca0da99a5fe/utils/run-lifecycle/run-lifecycle.js#L85) the `pkg` we are passing in is an internal [`Package`](https://github.com/lerna/lerna/blob/9f5c824b2f15119c10003dcc709f3ca0da99a5fe/core/package/index.js#L33) instance, _not_ the actual package.json's `JSON` object. And instances of that class have a _lot_ of non-enumerable fields and such.

So you're totally right, this is our bug, and I'mma fix it. Thanks for your persistence!

**Comment by busticated** [comment]
> Thanks for being patient with me

likewise 🙏 👍 

i can't say it enough: super-big thanks for all the work you and your fellow maintainers do here ❤️

**Comment by evocateur** [comment]
Published in [v3.9.0](https://github.com/lerna/lerna/releases/tag/v3.9.0)
